### ISSUE TITLE
Using Walrus Operator to bind Generic Parametrization

### ISSUE THREAD

**Issue by randolf-scholz** [python/mypy#13922]
**Bug Report**

Not 100% sure if this is really a bug, or behaviour outside of the specification of the Walrus Operator...

**To Reproduce**

```python
from typing import Generic, Literal, TypeAlias, TypeVar
from abc import ABC, abstractmethod

T = TypeVar("T")

class Foo(ABC, Generic[T]):
    @abstractmethod
    def f(self, x: T) -> str:
        ...

class Bar(Foo[KEY := Literal["A", "B", "C"]]):  # ✘ raises [misc, valid-type]
    assert KEY == Literal["A", "B", "C"]  # ✘  Cannot determine type of "KEY"  [has-type]

    def f(self, x: KEY) -> str:
        return f"The key is {x=}."

print(Bar().f("A"))    # ✔
print(Bar().f("XXX"))  # ✘ no error raised
```

**Expected Behavior**

At runtime, this code works fine - the assert statement holds. So I would have expected `mypy` to recognize `KEY` as a valid type. Of course, in practice it seems the better alternative is to simply use something along the lines of

```python
KEY: TypeAlias = Literal["A", "B", "C"]

class Bar(Foo[KEY]):
    def f(self, x: KEY) -> str:
        return f"The key is {x=}."
```

**Actual Behavior**

<details>
<summary>What went wrong? Paste mypy's output.</summary>

```
bugs/typing/walrus_generic.py:13:11: error: Type expected within [...]  [misc]
    class Bar(Foo[KEY := Literal["A", "B", "C"]]):  # ✘  raises [misc, valid-type]
              ^
bugs/typing/walrus_generic.py:13:11: error: Invalid base class "Foo"  [misc]
    class Bar(Foo[KEY := Literal["A", "B", "C"]]):  # ✘  raises [misc, valid-type]
              ^
bugs/typing/walrus_generic.py:14:12: error: Cannot determine type of "KEY"  [has-type]
        assert KEY == Literal["A", "B", "C"]
               ^~~
bugs/typing/walrus_generic.py:15:20: error: Variable "walrus_generic.KEY" is not valid as a type  [valid-type]
        def f(self, x: KEY) -> str:
                       ^
bugs/typing/walrus_generic.py:15:20: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
Found 4 errors in 1 file (checked 1 source file)
```

</details>

**Your Environment**

- Mypy version used: 0.982
- Python version used: 3.10.6

**Comment by erictraut** [comment]
The walrus operator isn't allowed within a type expression. Nor are variables. Mypy is correct to report errors here. Recommend closing.

**Comment by AlexWaygood** [comment]
Yeah, this would require a change to the typing spec (and I don't think there's a strong argument for doing so), and that would need to be discussed at https://github.com/python/typing
